java/ql/lib/semmle/code/java/dataflow/ExternalFlowConfiguration.qll

```diff
@@ -0,0 +1,95 @@
+/**
+ * This module provides extensible predicates for configuring which kinds of MaD models
+ * are applicable to a given query.
+ */
+
+private import ExternalFlowExtensions
+
+/**
+ * Holds if the specified kind of source model is supported for the current query.
+ */
+extensible private predicate supportedThreatModel(string kind);
+
+/**
+ * Holds if the specified kind of source model is containted within the specified group.
+ */
+extensible predicate threatModelGrouping(string kind, string group);
+
+/**
+ * Finds all of the threat models that are ancestors of the specified kind.
+ */
+private string parentThreatModel(string kind) {
+  exists(string parent | threatModelGrouping(kind, parent) |
+    result = parent or result = parentThreatModel(parent)
+  )
+}
+
+/**
+ * Finds all of the threat models that are descendants of the specified kind/group.
+ */
+private string childThreatModel(string group) {
+  exists(string child | threatModelGrouping(child, group) |
+    result = child or result = childThreatModel(child)
+  )
+}
+
+/**
+ * Holds if source models of the specified kind are
+ * supported for the current query.
+ */
+bindingset[kind]
+predicate supportedSourceModel(string kind) {
+  // expansive threat model includes all kinds
+  supportedThreatModel("expansive")
+  or
+  // check if this kind is supported directly
+  supportedThreatModel(kind)
+  or
+  // check if one of this kind's ancestors are supported
+  exists(string group | group = parentThreatModel(kind) | supportedThreatModel(group))
+  or
+  // if supportedThreatModel is empty, check if kind is a subtype of "standard"
+  not supportedThreatModel(_) and
+  ("standard" = parentThreatModel(kind) or "standard" = kind)
+}
+
+private string getGlobalGroups() { result = ["standard", "expansive"] }
+
+/**
+ * A class that represents a kind of any model or group.
+ */
+private class Kind extends string {
+  Kind() {
+    sourceModel(_, _, _, _, _, _, _, this, _) or
+    sinkModel(_, _, _, _, _, _, _, this, _) or
+    summaryModel(_, _, _, _, _, _, _, _, this, _) or
+    experimentalSourceModel(_, _, _, _, _, _, _, this, _, _) or
+    experimentalSinkModel(_, _, _, _, _, _, _, this, _, _) or
+    experimentalSummaryModel(_, _, _, _, _, _, _, _, this, _, _) or
+    supportedThreatModel(this) or
+    threatModelGrouping(this, _) or
+    threatModelGrouping(_, this) or
+    this = getGlobalGroups()
+  }
+}
+
+/**
+ * Gets the related source model kind(s) under the specified threat model.
+ */
+string relatedSourceModel(Kind kind) {
+  // Use the kinds provided by the query if the standard threat model is enabled or if no threat models are provided.
+  (supportedThreatModel("standard") or not supportedThreatModel(_)) and
+  result = kind
```

Comment:
Trying to wrap my head around this. Does this mean that when `supportedThreatModel` is `["standard"]` or empty, any `kind` that is explicitly passed in here will be _related_?

Reply:
Yes, that is correct.
Typically, define the set of flow sources by extending an abstract class (this is how it is done both in Java and C#).
In `FlowSources.qll` in Java we have
```codeql
private class ExternalRemoteFlowSource extends RemoteFlowSource {
  ExternalRemoteFlowSource() { sourceNode(this, "remote") }

  override string getSourceType() { result = "external" }
}
```
This means that all sources that has kind `remote` are included as remote flow sources.
However, if a query needs to add *extra* sources then locally to the query one could extend the abstract `RemoteFlowSource` class like this
```codeql
private class MyQuerySpecificSources extends RemoteFlowSource {
  MyQuerySpecificSources() { sourceNode(this, "mykind") }
  ...
}
```
If this query then uses the `RemoteFlowSource` class as the sources for the query (which is common programming pattern for all existing queries) then all sources of kind "remote" or "mykind" would be used - and these are the ones that I call *hardcoded* kinds.
An example of production code from `XmlInjection.ql`:
```
/**
 * A taint-tracking configuration for untrusted user input used in XML.
 */
module XmlInjectionConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
 ...
}

That is, most queries has a *hardcoded* set (that might be query specific) or source kinds that are used for a specific query.

The idea with the `relatedSourceModel` predicate is that if no supported threat model is given or if the threat model is set to standard then we use *exactly* the kinds that are hardcoded in the query and this can be achieved by related each kind to itself (lets discuss in person if this is unclear).